shadowsocks/connectivity.go

```diff
@@ -0,0 +1,115 @@
+	tcpTimeoutMs        = udpTimeoutMs * udpMaxRetryAttempts
+	udpTimeoutMs        = 1000
+	udpMaxRetryAttempts = 5
+	bufferLength        = 512
+)
+
+// ConnectivityResult holds the result of a connectivity check.
+type ConnectivityResult struct {
+	// Whehter the proxy is reachable through TCP.
+	IsReachable bool
+	// Whether the client's authentication credentials are valid.
+	IsAuthenticated bool
+	// Whether the proxy supports UDP forwarding.
+	IsUDPSupported bool
+}
+
+// CheckConnectivity determines whether the Shadowsocks proxy associated to `client` can relay TCP
+// and UDP traffic under the current network.
+func CheckConnectivity(client shadowsocks.Client) *ConnectivityResult {
+	tcpChan := make(chan error, 1)
+	udpChan := make(chan error, 1)
+	// Check whether the proxy is reachable and that the client is able to authenticate to the proxy
+	go func() {
+		tcpChan <- checkTCPConnectivityWithHTTP(client, "example.com")
+	}()
+	// Check whether UDP is supported
+	go func() {
+		udpChan <- CheckUDPConnectivityWithDNS(client, shadowsocks.NewAddr("1.1.1.1:53", "udp"))
+	}()
+
+	result := &ConnectivityResult{IsReachable: true, IsAuthenticated: true}
+	tcpErr := <-tcpChan
+	if tcpErr != nil {
+		// Determine whether the proxy was reachable over TCP based on the error type; reachability
+		// errors occur when dialing to the proxy, authentication errors occur during reads.
+		_, isDialError := tcpErr.(*net.OpError)
+		result.IsReachable = !isDialError
+		result.IsAuthenticated = false
+	}
+	result.IsUDPSupported = <-udpChan == nil
+	return result
+}
+
+// CheckUDPConnectivityWithDNS determines whether the Shadowsocks proxy represented by `client` and
+// the network support UDP traffic by issuing a DNS query though a resolver at `resolverAddr`.
+// Returns nil on success or an error on failure.
+func CheckUDPConnectivityWithDNS(client shadowsocks.Client, resolverAddr net.Addr) error {
+	conn, err := client.ListenUDP(nil)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	conn.SetReadDeadline(time.Now().Add(time.Millisecond * udpTimeoutMs))
+	buf := make([]byte, bufferLength)
+	for attempt := 0; attempt < udpMaxRetryAttempts; attempt++ {
+		_, err := conn.WriteTo(getDNSRequest(), resolverAddr)
+		if err != nil {
+			continue
+		}
+		n, _, err := conn.ReadFrom(buf)
+		if n == 0 && err != nil {
+			continue
+		}
+		return nil
+	}
+	return errors.New("UDP not supported")
+}
+
+// checkTCPConnectivityWithHTTP determines whether the proxy is reachable over TCP and validates the
+// client's authentication credentials by performing an HTTP HEAD request to `targetDomain`.
+func checkTCPConnectivityWithHTTP(client shadowsocks.Client, targetDomain string) (err error) {
+	conn, err := client.DialTCP(nil, net.JoinHostPort(targetDomain, "80"))
+	if err != nil {
+		return
+	}
+	defer conn.Close()
+	conn.SetReadDeadline(time.Now().Add(time.Millisecond * tcpTimeoutMs))
+	payload := fmt.Sprintf("HEAD / HTTP/1.1\r\nHost: %v\r\n\r\n", targetDomain)
```

Comment:
To escape, you can use `net.http.Header`. There's a `Add` method. Then the `Write()` method will return the wire format.

Reply:
Done - I'm now using the net/http package to write the  request.

Reply:
Nice!